Add HTTPS Support

jbarnett-r7 opened the pull request:
This PR adds HTTPS support to the server and client for using a remote data service.

A few things to note:
* I only tested with basic setup on everything. Testing with fancier encryption algorithms and such would be beneficial.
* There is a process crashing bug when you attempt to use HTTP to connect to a HTTPS data service. This bug actually hits whenever you make a call that results in a bad HTTP request. We need to fix that but will do it in a different PR.
* I added some help text and flags to the `add_data_service` and `msfdb` commands that should be given a once over for clarity and typos. 

## Verification

- [x] Generate an SSL key and cert file with .pem extensions.
- [x] Start the server with SSL enabled `msfdb -s -c <path to cert file> -k <path to key file>`
- [x] Use a browser to verify the server is running correctly and using the correct protocol and cert.
- [x] Start `msfconsole`
- [x] Add an HTTPS data server `add_data_service -s -c <path to cert file> -p <port> <IP address>`
- [x] Verify you can connect to the secure data service and can create and retrieve data records.
- [x] Run a quick regression test to make sure everything still works with regular HTTP.

jbarnett-r7 committed:
WIP: HTTP server launching with HTTPS

jbarnett-r7 committed:
WIP: Enable HTTPS client.

Removed RemoteServiceEndpoint and using URI instead.

jbarnett-r7 committed:
Remove require statements for remote_service_endpoint

jbarnett-r7 committed:
Add option handling to msfdb

Can now specify custom interface and port.
Now able to specify starting in HTTPS mode.

jbarnett-r7 committed:
Merge branch 'goliath' into add_https

jbarnett-r7 committed:
Merge branch 'goliath' into add_https

jbarnett-r7 committed:
Add help text to cmd_add_data_service

jbarnett-r7 committed:
Add error handling when no data returned from server

mkienow-r7 commented on lib/metasploit/framework/data_service/remote/http/core.rb:
```diff
@@ -239,7 +242,10 @@ def build_client_pool(size)
-      http.use_ssl = true if @endpoint.use_ssl
+      if @endpoint.is_a?(URI::HTTPS)
+        http.use_ssl = true
+        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
```
Should `http.verify_mode` be set `OpenSSL::SSL::VERIFY_PEER`? Without that no verification will occur, correct? If the user provides a cert and key we need to assign those to `http.cert` and `http.key`; otherwise, assume they are using a cert that can be verified through the CA certificates installed on their system by default. Another option would be to support custom CA certs via `http.cert_store` and another parameter.
See [net-http-cheat-sheet/ssl_and_https](https://github.com/augustl/net-http-cheat-sheet/blob/master/ssl_and_https.rb)

jbarnett-r7 replied:
Yeah, I wasn't really sure how to get around this. Obviously for my testing I only had a self-signed cert, so this option was necessary. One thing I was thinking about doing was enabling cert verification by defaults adding a flag to the command to override it. What do you think about that?

In the meantime I'll check into the things you suggested. I wasn't aware of all of them.

mkienow-r7 replied:
Wouldn't verification of self-signed certificates work without issue if you provided the cert and key to the server/client side?

mkienow-r7 commented on lib/msf/core/db_manager/http/http_db_manager_service.rb:
```diff
@@ -16,6 +16,15 @@ def start(opts)
+      ssl_opts = {}
+      ssl_opts[:private_key_file] = opts[:ssl_key]
+      ssl_opts[:cert_chain_file] = opts[:ssl_cert]
+      ssl_opts[:verify_peer] = false
```
Should `:verify_peer` be set `true`?

jbarnett-r7 replied:
I think this goes with the above comment. Will adjust functionality based on what we figure out there.

jbarnett-r7 replied:
I just tested it with the server set to `true` and no changes on the client and it still works. I'm not  sure that this option actually works?

mkienow-r7 commented on lib/msf/ui/console/command_dispatcher/db.rb:
```diff
@@ -100,21 +99,43 @@ def cmd_list_data_services()
+            protocol = "http"
             while (arg = args.shift)
               case arg
+                when '--help'
```
The help flag is inconsistent with other commands. Perhaps we should make `-h` and `--help` aliases for help. Making the hostname/address a positional argument would be consistent with a number of the other commands.

jbarnett-r7 replied:
Yeah I'm ok with this. I'll make those changes.

mkienow-r7 commented on msfdb:
```diff
@@ -7,4 +7,76 @@
+    opts[:ssl_key] = p
+  end
+
+  opt.on_tail('-h', '--help', 'Show this message') do
```
Modify description to "Show this help information" to be consistent with other commands.

mkienow-r7 commented on msfdb:
```diff
@@ -7,4 +7,76 @@
+    opts[:ssl_cert] = p
+  end
+
+  opt.on('-k', '--key       <path/to/key.pem>', String, 'Path to your SSL Key file') do |p|
```
Minor: we could probably drop "your"

mkienow-r7 commented on msfdb:
```diff
@@ -7,4 +7,76 @@
+    opts[:ssl] = true
+  end
+
+  opt.on('-c', '--cert      <path/to/cert.pem>', String, 'Path to your SSL Certificate file') do |p|
```
Minor: we could probably drop "your"

jbarnett-r7 committed:
Merge branch 'goliath' into add_https

jbarnett-r7 committed:
Address PR comments regarding command options

jbarnett-r7 committed:
Fix bug with -s flag

jbarnett-r7 committed:
Add verification methods to HTTPS

This commit enables peer verification for SSL.
It also gives the user options to verify the server if the server uses a self-signed cert.
There is an override to skip verification as well.

jbarnett-r7 commented:
@mkienow-r7 I think this most recent commit should address most of your concerns. Please let me know how it works.

mkienow-r7 commented on lib/msf/ui/console/command_dispatcher/db.rb:
```diff
@@ -100,21 +99,50 @@ def cmd_list_data_services()
                   port = args.shift
+                when '-s', '--ssl'
+                  protocol = "https"
+                when '-c'
```
Would be nice to include `--cert` to match `msfdb` options

jbarnett-r7 replied:
I had that, and then took it off. I don't know if this paradigm is used elsewhere in the app, but I was going with a certain style where if you pass a value after the flag then it would only have the `-letter` option. If the flag is just a setting or toggle, then it would have a `--word` option.

Does this make sense? Or is it a convention that I just created in my head?

mkienow-r7 replied:
The `hosts` command has both `-letter` and `--word` options for options that take a value.
```
-S,--search       Search string to filter by
-i,--info         Change the info of a host
```

jbarnett-r7 replied:
OK, so I guess I can add it back to conform with what the rest of the app is doing.

mkienow-r7 commented:
If the private key is encrypted when it is created, for example, one didn't pass `-nodes` to `openssl` there will be problems. Every request sent to the `msfdb` remote side will result in the user being prompted to `Enter PEM pass phrase`. I'm fine integrating this as-is as long as we get a backlog item to support encrypted private keys.

jbarnett-r7 committed:
Add long option for cert.

mkienow-r7 committed:
Merge branch 'goliath' into add_https
